<h3>NLP Tutorial: Text Classification Using FastText</h3>

##### Dataset Credits: https://www.kaggle.com/datasets/saurabhshahane/ecommerce-text-classification

We have a dataset of ecommerce item description. Total 4 categories,
1. Household
2. Electronics
3. Clothing and Accessories
4. Books

The task at hand is to classify a product into one of the above 4 categories based on the product description

In [1]:
import pandas as pd

df= pd.read_csv("ecommerce_dataset.csv", names=["category", "description"], header=None)
print(df.shape)
df.head(3)

C:\Users\abhinav.aggarwal\AppData\Local\Temp\ipykernel_13676\3681045313.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


(50425, 2)


category                                        description
0  Household  Paper Plane Design Framed Wall Hanging Motivat...
1  Household  SAF 'Floral' Framed Painting (Wood, 30 inch x ...
2  Household  SAF 'UV Textured Modern Art Print Framed' Pain...

**Drop NA values**

In [2]:
df.dropna(inplace=True)
df.shape

(50424, 2)

In [3]:
df.category.unique()

array(['Household', 'Books', 'Clothing & Accessories', 'Electronics'],
      dtype=object)

In [4]:
df.category.replace("Clothing & Accessories", "Clothing_Accessories", inplace=True)

C:\Users\abhinav.aggarwal\AppData\Local\Temp\ipykernel_13676\2912538940.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df.category.replace("Clothing & Accessories", "Clothing_Accessories", inplace=True)


In [5]:
df.category.unique()

array(['Household', 'Books', 'Clothing_Accessories', 'Electronics'],
      dtype=object)

When you train a fasttext model, it expects labels to be specified with __label__ prefix. We will just create a third column in the dataframe that has __label__ as well as the product description

In [6]:
df['category'] = '__label__' + df['category'].astype(str)
df.head(5)

category                                        description
0  __label__Household  Paper Plane Design Framed Wall Hanging Motivat...
1  __label__Household  SAF 'Floral' Framed Painting (Wood, 30 inch x ...
2  __label__Household  SAF 'UV Textured Modern Art Print Framed' Pain...
3  __label__Household  SAF Flower Print Framed Painting (Synthetic, 1...
4  __label__Household  Incredible Gifts India Wooden Happy Birthday U...

In [7]:
df['category_description'] = df['category'] + ' ' + df['description']
df.head(3)

category                                        description  \
0  __label__Household  Paper Plane Design Framed Wall Hanging Motivat...   
1  __label__Household  SAF 'Floral' Framed Painting (Wood, 30 inch x ...   
2  __label__Household  SAF 'UV Textured Modern Art Print Framed' Pain...   

                                category_description  
0  __label__Household Paper Plane Design Framed W...  
1  __label__Household SAF 'Floral' Framed Paintin...  
2  __label__Household SAF 'UV Textured Modern Art...

**Pre-procesing**
1. Remove punctuation
2. Remove extra space
3. Make the entire sentence lower case

In [8]:
import re

text = "  VIKI's | Bookcase/Bookshelf (3-Shelf/Shelve, White) | ? . hi"
text = re.sub(r'[^\w\s\']',' ', text)
text = re.sub(' +', ' ', text)
text.strip().lower()

"viki's bookcase bookshelf 3 shelf shelve white hi"

In [9]:
def preprocess(text):
    text = re.sub(r'[^\w\s\']',' ', text)
    text = re.sub(' +', ' ', text)
    return text.strip().lower() 

In [10]:
df['category_description'] = df['category_description'].map(preprocess)
df.head()

category                                        description  \
0  __label__Household  Paper Plane Design Framed Wall Hanging Motivat...   
1  __label__Household  SAF 'Floral' Framed Painting (Wood, 30 inch x ...   
2  __label__Household  SAF 'UV Textured Modern Art Print Framed' Pain...   
3  __label__Household  SAF Flower Print Framed Painting (Synthetic, 1...   
4  __label__Household  Incredible Gifts India Wooden Happy Birthday U...   

                                category_description  
0  __label__household paper plane design framed w...  
1  __label__household saf 'floral' framed paintin...  
2  __label__household saf 'uv textured modern art...  
3  __label__household saf flower print framed pai...  
4  __label__household incredible gifts india wood...

**Train Test Split**

In [11]:
from sklearn.model_selection import train_test_split

train, test = train_test_split(df, test_size=0.2)

In [12]:
train.shape, test.shape

((40339, 3), (10085, 3))

In [13]:
train.to_csv("ecommerce.train", columns=["category_description"], index=False, header=False)
test.to_csv("ecommerce.test", columns=["category_description"], index=False, header=False)

**Train the model and evaluate performance**

In [14]:
import fasttext

model = fasttext.train_supervised(input="ecommerce.train")
model.test("ecommerce.test")

(10084, 0.9691590638635462, 0.9691590638635462)

First parameter (10084) is test size. Second and third parameters are precision and recall respectively. You can see we are getting around 96% precision which is pretty good

**Now let's do prediction for few product descriptions**

In [15]:
model.predict("wintech assemble desktop pc cpu 500 gb sata hdd 4 gb ram intel c2d processor 3")

(('__label__electronics',), array([0.99610347]))

In [16]:
model.predict("ockey men's cotton t shirt fabric details 80 cotton 20 polyester super combed cotton rich fabric")

(('__label__clothing_accessories',), array([1.00001001]))

In [17]:
model.predict("think and grow rich deluxe edition")

(('__label__books',), array([1.00000978]))

In [18]:
model.get_nearest_neighbors("painting")

[(0.9987408518791199, 'descriptionsure'),
 (0.9987408518791199, 'fumed'),
 (0.9987408518791199, 'sellerthe'),
 (0.9987408518791199, 'diant'),
 (0.9987396597862244, 'sham'),
 (0.9987346529960632, '1sq'),
 (0.9987264275550842, 'steelproduct'),
 (0.9987232089042664, 'coatingit'),
 (0.9987232089042664, 'illation'),
 (0.9987232089042664, 'lopper')]

In [19]:
model.get_nearest_neighbors("sony")

[(0.9993534684181213, 'swifter'),
 (0.9993500709533691, 'mx130discrete'),
 (0.9993500709533691, 'cd0051tx'),
 (0.9993500709533691, 'olufsen'),
 (0.9993484020233154, '5atm'),
 (0.9993482828140259, 'accelerometers'),
 (0.9993482828140259, 'ios9'),
 (0.9993482828140259, '20gm'),
 (0.9993482828140259, 'mifit'),
 (0.9993482828140259, '247mm')]

In [20]:
model.get_nearest_neighbors("banglore")

[(0.0, 'to'),
 (0.0, 'and'),
 (0.0, 'a'),
 (0.0, 'with'),
 (0.0, 'for'),
 (0.0, 'is'),
 (0.0, '</s>'),
 (0.0, 'shopmetro'),
 (0.0, '111101_multicolor'),
 (0.0, 'frayn')]